File: EaglerTools/GenerateBlocks.py
import os
import shutil
import json

base_path = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

source_folder = os.path.join(base_path,"EaglerTools", "makeBlock")
block_texture_folder = os.path.join(base_path, "desktopRuntime", "resources", "assets", "starlike", "textures", "blocks")
block_model_folder = os.path.join(base_path, "desktopRuntime", "resources", "assets", "starlike", "models", "block")
item_model_folder = os.path.join(base_path, "desktopRuntime", "resources", "assets", "starlike", "models", "item")
blockstate_folder = os.path.join(base_path,"desktopRuntime", "resources", "assets","starlike", "blockstates")
lang_file_path = os.path.join(base_path,"desktopRuntime", "resources", "assets","starlike", "lang", "en_US.lang")

os.makedirs(block_texture_folder, exist_ok=True)
os.makedirs(block_model_folder, exist_ok=True)
os.makedirs(item_model_folder, exist_ok=True)
os.makedirs(blockstate_folder, exist_ok=True)


items_starlike_java_path = os.path.join(
    base_path, "src", "main", "java", "net", "starlikeclient", "minecraft", "init", "ItemsStarlike.java"
)
blocks_java_path = os.path.join(base_path,"src", "game", "java", "net", "minecraft", "init", "Blocks.java")

blockId = 695; #replace with next id
def insert_in_file(file_path, search_text, insert_text, insert_after=True, tab_before=False):
    try:
        with open(file_path, 'r') as file:
            lines = file.readlines()
    except Exception as e:
        print(f"Error reading {file_path}: {e}")
        return False
    
    for i, line in enumerate(lines):
        if search_text in line:
            insert_line = insert_text + '\n'
            if tab_before:
                insert_line = '\t' + insert_line
            if insert_after:
                lines.insert(i + 1, insert_line)
            else:
                lines.insert(i, insert_line)
            break

    try:
        with open(file_path, 'w') as file:
            file.writelines(lines)
        print(f"Updated {file_path} successfully.")
        return True
    except Exception as e:
        print(f"Error writing to {file_path}: {e}")
        return False

def append_to_file(file_path, new_line):
    try:
        with open(file_path, 'a') as file:
            file.write(new_line + '\n')
        print(f"Appended to {file_path} successfully.")
    except Exception as e:
        print(f"Error appending to {file_path}: {e}")

print("Welcome to GenerateBlocks (EaglerTools)")
print("Drag the texture of the FULL block into the ")
for filename in os.listdir(source_folder):
    if filename.endswith(".png"):
        
        block_name = os.path.splitext(filename)[0]

        source_image = os.path.join(source_folder, filename)
        dest_image = os.path.join(block_texture_folder, filename)
        
        try:
            shutil.move(source_image, dest_image)
            print(f"Moved image: {filename} to {dest_image}")
        except Exception as e:
            print(f"Error moving file {filename}: {e}")
            continue

        blockstate_data = {
            "parent": "block/cube_all",
            "textures": {
                "all": f"starlike:blocks/{block_name}"
            }
        }
        blockstate_path = os.path.join(block_model_folder, f"{block_name}.json")
        try:
            with open(blockstate_path, 'w') as blockstate_file:
                json.dump(blockstate_data, blockstate_file, indent=4)
                print(f"Created blockstate JSON: {blockstate_path}")
        except Exception as e:
            print(f"Error writing blockstate JSON for {block_name}: {e}")
            continue

        item_model_data = {
            "parent": f"starlike:block/{block_name}",
            "display": {
                "thirdperson": {
                    "rotation": [10, -45, 170],
                    "translation": [0, 1.5, -2.75],
                    "scale": [0.375, 0.375, 0.375]
                }
            }
        }
        item_model_path = os.path.join(item_model_folder, f"{block_name}.json")
        try:
            with open(item_model_path, 'w') as item_model_file:
                json.dump(item_model_data, item_model_file, indent=4)
                print(f"Created item model JSON: {item_model_path}")
        except Exception as e:
            print(f"Error writing item model JSON for {block_name}: {e}")
            continue

        blockstate_json_data = {
            "variants": {
                "normal": {
                    "model": f"starlike:{block_name}"
                }
            }
        }
        blockstate_json_path = os.path.join(blockstate_folder, f"{block_name}.json")
        try:
            with open(blockstate_json_path, 'w') as blockstate_json_file:
                json.dump(blockstate_json_data, blockstate_json_file, indent=4)
                print(f"Created blockstate JSON: {blockstate_json_path}")
        except Exception as e:
            print(f"Error writing blockstate JSON for {block_name}: {e}")
            continue

        material_type = "Material.wood"
        hardness = "0.5f"
        resistance = "5.0f"
        soundType = "soundTypePiston"
        if any(keyword in block_name for keyword in ["log", "planks", "wood"]):
            material_type = "Material.wood"
            hardness = "0.5f"
            resistance = "10.0f"
            soundType = "soundTypeWood"
        elif any(keyword in block_name for keyword in ["steel", "iron", "grate", "bar", "copper", "gold"]):
            material_type = "Material.iron"
            hardness = "5f"
            resistance = "10.0f"
            soundType = "soundTypeMetal"
        elif any(keyword in block_name for keyword in ["dirt", "grass"]):
            material_type = "Material.ground"
            hardness = "0.5f"
            resistance = "1f"
            soundType = "soundTypeGround"    
        else:
            material_type = "Material.rock"
            hardness = "1.5f"
            resistance = "10.0f"
            soundType = "soundTypePiston"

        registration_line = (
            f"Block.registerBlock({blockId}, \"starlike:{block_name}\", "
            f"(new Block({material_type}, MapColor.purpleColor)).setHardness({hardness}).setResistance({resistance}) "
            f".setStepSound(Block.{soundType}).setUnlocalizedName(\"{block_name}\").setCreativeTab(CreativeTabs.tabStarlike));"
        )
        insert_in_file(items_starlike_java_path, "//autogenerateequals", registration_line)

        block_declaration = f"    public static Block {block_name};"
        get_registered_line = f"        {block_name} = getRegisteredBlock(\"starlike:{block_name}\");"
        insert_in_file(blocks_java_path, "//autogeneratevar", block_declaration)
        insert_in_file(blocks_java_path, "//autogenerateequals", get_registered_line)

        item_block_registration = f"Item.registerItemBlock(Blocks.{block_name});"
        insert_in_file(items_starlike_java_path, "//autogenerateregisterblock", item_block_registration, tab_before=True)
        item_block_render_registration = f'e.registerBlock(Blocks.{block_name},"starlike:{block_name}");'
        insert_in_file(items_starlike_java_path, "//autogeneraterender", item_block_render_registration, tab_before=True)
        lang_entry = f"tile.{block_name}.name={block_name.replace('_', ' ').capitalize()}"
        append_to_file(lang_file_path, lang_entry)
        blockId += 1

File: EaglerTools/GenerateBlocks.mjs
import path from 'path';
import fs from 'fs/promises';

let blockId = 695; // replace with next id

const basePath = path.dirname(path.dirname(path.resolve(import.meta.filename)));
const sourceFolder = path.join(basePath, 'EaglerTools', 'makeBlock');
const starlikeAssetsFolder = path.join(
  basePath,
  'desktopRuntime',
  'resources',
  'assets',
  'starlike',
);

const blockTextureFolder = path.join(
  starlikeAssetsFolder,
  'textures',
  'blocks',
);
const blockModelFolder = path.join(starlikeAssetsFolder, 'models', 'block');
const itemModelFolder = path.join(starlikeAssetsFolder, 'models', 'item');
const blockstateFolder = path.join(starlikeAssetsFolder, 'blockstates');
const langFilePath = path.join(starlikeAssetsFolder, 'lang', 'en_US.lang');

const itemsStarlikeJavaPath = path.join(
  basePath,
  'src',
  'main',
  'java',
  'net',
  'starlikeclient',
  'minecraft',
  'init',
  'ItemsStarlike.java',
);
const blocksJavaPath = path.join(
  basePath,
  'src',
  'game',
  'java',
  'net',
  'minecraft',
  'init',
  'Blocks.java',
);

async function createDirectories() {
  await fs.mkdir(blockTextureFolder, { recursive: true });
  await fs.mkdir(blockModelFolder, { recursive: true });
  await fs.mkdir(itemModelFolder, { recursive: true });
  await fs.mkdir(blockstateFolder, { recursive: true });
}

async function insertInFile(
  filePath,
  searchText,
  insertText,
  insertAfter = true,
  tabBefore = false,
) {
  try {
    const content = await fs.readFile(filePath, 'utf8');
    const lines = content.split('\n');

    for (let i = 0; i < lines.length; i++) {
      if (lines[i].includes(searchText)) {
        let insertLine = insertText + '\n';
        if (tabBefore) {
          insertLine = '\t' + insertLine;
        }
        if (insertAfter) {
          lines.splice(i + 1, 0, insertLine);
        } else {
          lines.splice(i, 0, insertLine);
        }
        break;
      }
    }

    await fs.writeFile(filePath, lines.join('\n'));
    console.log(`Updated ${filePath} successfully.`);
    return true;
  } catch (err) {
    console.error(`Error with file ${filePath}: ${err}`);
    return false;
  }
}

async function appendToFile(filePath, newLine) {
  try {
    await fs.appendFile(filePath, newLine + '\n');
    console.log(`Appended to ${filePath} successfully.`);
  } catch (err) {
    console.error(`Error appending to ${filePath}: ${err}`);
  }
}

async function processBlocks() {
  console.log('Welcome to GenerateBlocks (EaglerTools)');
  console.log(
    'Drag the texture of the FULL block into the EaglerTools/makeBlock folder',
  );

  try {
    const files = await fs.readdir(sourceFolder);

    for (const filename of files) {
      if (!filename.endsWith('.png')) continue;

      const blockName = path.parse(filename).name;
      const sourceImage = path.join(sourceFolder, filename);
      const destImage = path.join(blockTextureFolder, filename);

      try {
        await fs.rename(sourceImage, destImage);
        console.log(`Moved image: ${filename} to ${destImage}`);

        // Create blockstate data
        const blockModelData = {
          parent: 'block/cube_all',
          textures: {
            all: `starlike:blocks/${blockName}`,
          },
        };

        // Create item model data
        const itemModelData = {
          parent: `starlike:block/${blockName}`,
          display: {
            thirdperson: {
              rotation: [10, -45, 170],
              translation: [0, 1.5, -2.75],
              scale: [0.375, 0.375, 0.375],
            },
          },
        };

        // Create blockstate JSON data
        const blockstateData = {
          variants: {
            normal: {
              model: `starlike:${blockName}`,
            },
          },
        };

        // Write JSON files
        await fs.writeFile(
          path.join(blockModelFolder, `${blockName}.json`),
          JSON.stringify(blockModelData, null, 4),
        );

        await fs.writeFile(
          path.join(itemModelFolder, `${blockName}.json`),
          JSON.stringify(itemModelData, null, 4),
        );

        await fs.writeFile(
          path.join(blockstateFolder, `${blockName}.json`),
          JSON.stringify(blockstateData, null, 4),
        );

        // Determine material properties
        let materialType = 'Material.rock';
        let hardness = '1.5f';
        let resistance = '10.0f';
        let soundType = 'soundTypePiston';

        if (/log|planks|wood/.test(blockName)) {
          materialType = 'Material.wood';
          hardness = '0.5f';
          resistance = '10.0f';
          soundType = 'soundTypeWood';
        } else if (/steel|iron|grate|bar|copper|gold/.test(blockName)) {
          materialType = 'Material.iron';
          hardness = '5f';
          resistance = '10.0f';
          soundType = 'soundTypeMetal';
        } else if (/dirt|grass/.test(blockName)) {
          materialType = 'Material.ground';
          hardness = '0.5f';
          resistance = '1f';
          soundType = 'soundTypeGround';
        }

        // Generate and insert Java code
        const registrationLine =
          `Block.registerBlock(${blockId}, "starlike:${blockName}", ` +
          `(new Block(${materialType}, MapColor.purpleColor)).setHardness(${hardness}).setResistance(${resistance}) ` +
          `.setStepSound(Block.${soundType}).setUnlocalizedName("${blockName}").setCreativeTab(CreativeTabs.tabStarlike));`;

        await insertInFile(
          itemsStarlikeJavaPath,
          '//autogenerateequals',
          registrationLine,
        );

        const blockDeclaration = `    public static Block ${blockName};`;
        const getRegisteredLine = `        ${blockName} = getRegisteredBlock("starlike:${blockName}");`;

        await insertInFile(
          blocksJavaPath,
          '//autogeneratevar',
          blockDeclaration,
        );
        await insertInFile(
          blocksJavaPath,
          '//autogenerateequals',
          getRegisteredLine,
        );

        const itemBlockRegistration = `Item.registerItemBlock(Blocks.${blockName});`;
        await insertInFile(
          itemsStarlikeJavaPath,
          '//autogenerateregisterblock',
          itemBlockRegistration,
          true,
          true,
        );

        const itemBlockRenderRegistration = `e.registerBlock(Blocks.${blockName},"starlike:${blockName}");`;
        await insertInFile(
          itemsStarlikeJavaPath,
          '//autogeneraterender',
          itemBlockRenderRegistration,
          true,
          true,
        );

        const langEntry = `tile.${blockName}.name=${
          blockName.replace('_', ' ').charAt(0).toUpperCase() +
          blockName.slice(1)
        }`;
        await appendToFile(langFilePath, langEntry);

        blockId++;
      } catch (err) {
        console.error(`Error processing ${filename}: ${err}`);
        continue;
      }
    }
  } catch (err) {
    console.error(`Error reading source folder: ${err}`);
  }
}

// Run the script
createDirectories()
  .then(() => processBlocks())
  .catch((err) => console.error('Error:', err));
